docs: update hold-tap page (#2888)
zmkfirmware/zmk: docs/docs/keymaps/behaviors/hold-tap.mdx

```dts title="layer-tap"
#include <dt-bindings/zmk/behaviors.h>

/ {
    behaviors {
        lt: layer_tap {
            compatible = "zmk,behavior-hold-tap";
            #binding-cells = <2>;
            flavor = "tap-preferred";
            tapping-term-ms = <200>;
            bindings = <&mo>, <&kp>;
            display-name = "Layer-Tap";
        };
    };
};

```

If you copy-paste these as-is, you will overwrite the predefined behaviors.
Name and label them something else, e.g. `mt2: mod_tap_2`, to avoid this.
You'll then be able to use your new behavior with `&mt2`.

</TabItem>
<TabItem value="homerow_mods">

The most popular form of homerow mods is known as "timeless homerow mods", configured to minimise the dependency on timing.
Timeless homerow mods use both a "left hand" and a "right hand" behavior.

Below is an example of a left hand behavior:

```dts
\ {
    behaviors {
        hml: home_row_mod_left {
            compatible = "zmk,behavior-hold-tap";
            #binding-cells = <2>;
            flavor = "balanced";
            require-prior-idle-ms = <150>;
            tapping-term-ms = <280>;
            quick-tap-ms = <175>;
            bindings = <&kp>, <&kp>;

>             hold-trigger-key-positions = < ... >; // List of keys on the left hand side of the keyboard

Reviewer: ```suggestion
            hold-trigger-key-positions = < ... >; // List of keys on the right side of the keyboard
```

            hold-trigger-on-release;
        };

>     };

Reviewer: ```suggestion
        hmr: home_row_mod_right {
            compatible = "zmk,behavior-hold-tap";
            #binding-cells = <2>;
            flavor = "balanced";
            require-prior-idle-ms = <150>;
            tapping-term-ms = <280>;
            quick-tap-ms = <175>;
            bindings = <&kp>, <&kp>;
            hold-trigger-key-positions = < ... >; // List of keys on the left side of the keyboard
            hold-trigger-on-release;
        };
    };
```
If we are encouraging copy pasting, why not define both?

Author: Because they are identically barring hold-trigger-key-positions, which they need to fill out themselves anyway.

Reviewer: True, but I think the "use `&hml`/`&hmr`" remark on the other suggestion is necessary and not sure how to do that without defining `hmr` in some way.

};
```

> If you wish to use this configuration of homerow mods, you will need to define the corresponding left and right behaviors. Below is a brief overview of the options set.

Reviewer: ```suggestion
If you wish to use this configuration of homerow mods, you will need to define the corresponding left and right behaviors as above, then use `&hml` for homerow keys on the left side of the keyboard, and `&hmr` for ones on the right side.

Below is a brief overview of the options set.
```

- The ["balanced" flavor](#interrupt-flavors) produces a "hold" if another key is both pressed and released within the tapping-term.
  This interrupt flavor is primarily used to decide between hold and tap, rather than using [tapping-term-ms](#tapping-term-ms).
  This matches how modifiers are used a majority of the time (hold shift, tap a key, release shift).

> - [`require-prior-idle-ms`](#require-prior-idle-ms) is set to 125ms to resolve to taps immediately when typing at speed, which all but completely eliminates any typing delay.

Reviewer: ```suggestion
- [`require-prior-idle-ms`](#require-prior-idle-ms) is set to 125ms to resolve to taps immediately when typing at speed, which helps eliminate typing delay.
```

- [Positional hold taps](#positional-hold-tap-and-hold-trigger-key-positions) with [`hold-trigger-on-release`] are used to avoid accidental hold resolutions when typing sequences of letters all with the same hand.
- [`tapping-term-ms`](#tapping-term-ms) is set to a higher value of 280ms, but not unreasonably high, so that same hand modifiers can still be used with intention.

> - [`quick-tap-ms`](#quick-tap-ms) is set to 175ms, for the event where a key may wish to be held down.

Reviewer: ```suggestion
- [`quick-tap-ms`](#quick-tap-ms) is set to 175ms, for the event where a tapped key may wish to be held down.
```

> Depending on your typing behavior, you may wish to adjust this configuration somewhat. For example, adjusting the specific timings used, or switching the flavor to `tap-unless-interrupted`.

Reviewer: ```suggestion
Depending on your typing behavior, you may wish to adjust this configuration. For example, you can adjust the specific timings used, or define another set of hold-taps for more dextrous fingers like the index with different parameters.
```

I guess this is a bit subjective opinion territory but `tap-unless-interrupted` is niche and I am not sure it is worth mentioning here.

Author: I think it might be useful to mention the niche cases where it *can* come in useful. I feel like people often struggle to understand "I want this sort of tapping behavior -> Ok this particular flavor is ideal for that".

Reviewer: I was being polite, I don't think this flavor should be recommended given how counter-intuitive it is with the tapping term. (It is already documented in the flavors.)

Reviewer: OK, I disagree but I won't push it further.

Including `hold-trigger-key-positions` in your hold-tap definition turns on the positional hold-tap feature. With positional hold-tap enabled, if you press any key **NOT** listed in `hold-trigger-key-positions` before `tapping-term-ms` expires, it will produce a tap.

In all other situations, positional hold-tap will not modify the behavior of your hold-tap. Positional hold-tap is useful when used with home-row modifiers: for example, if you have a home-row modifier key in the left hand, by including only key positions from the right hand in `hold-trigger-key-positions`, you will only get hold behaviors during cross-hand key combinations.

:::info
Note that `hold-trigger-key-positions` is an array of key position indexes. Key positions are numbered sequentially according to your keymap, starting with 0. So if the first key in your keymap is Q, this key is in position 0. The next key (probably W) will be in position 1, et cetera.
:::

#### `hold-trigger-on-release`

If set, instead of the keys listed in `hold-trigger-key-positions` producing a tap when _pressed_ before `tapping-term-ms` expires, they instead produce a tap when _released_ before `tapping-term-ms` expires.

> #### Example

Author: I see your point, but I'm not personally a fan of using details boxes here. I think with the reordering of the config settings it should be fine.

> #### Example

Reviewer: I think they take up too much real estate and inflates their importance relative to other settings, just because they end up being documented for a full page height. I'd try to at least make them more concise in this case.

> #### Example

Author: That's fair. I'll shove them in details boxes.

For example, if the hold is `&kp LGUI` and the tap is `&sk LGUI`, then with `hold-while-undecided-linger` enabled, the host will see `LGUI` held down continuously until the sticky key is finished, instead of seeing a release and press when transitioning from hold to sticky key.

:::info[Alt/Win/Cmd behavior]
In some applications/desktop environments, pressing Alt keycodes by itself will have its own behavior like activate a menu and Gui keycodes will bring up the start menu or an application launcher.
:::

### `retro-tap`

If `retro-tap` is enabled, the tap behavior is triggered when releasing the hold-tap key if no other key was pressed in the meantime. The hold key does not activate until another key is pressed, meaning that it cannot be used for mouse events like Shift Click to select from your cursor position to mouse position.

For example, if you press `&mt LEFT_SHIFT A` and then release it without pressing another key, it will output `a`.

```dts
&mt {
    retro-tap;
};

> ```

Author: This is counter to your other suggestion of putting small example configurations under everything. I've added small examples under everything now, take a look and see what you think. I think we do all or nothing.

> ```

Reviewer: I was aware when I added this comment, I imagined the examples without `&mt {` part but that's fine too, thanks.

</Tabs>

## Custom Hold-Tap Configuration

### `tapping-term-ms`

This value defines how long a key must be pressed to trigger the "hold" behavior, alongside other factors described in [interrupt flavors](#interrupt-flavors). The default is 200ms.

```dts

> &my_mt{

Author: Mostly me copypasting different bits. Which one do you think would be preferable?

> &my_mt{

Reviewer: I think it's better to avoid referring to non-predefined behaviors if we can, so I'd prefer `&mt`.
